contracts/src/main/kotlin/com/contractsAndStates/states/RoadState.kt

```diff
@@ -45,3 +45,153 @@ data class RoadState(
+ * be a new player, the same player or null if no one meets the criteria.
+ */
+fun assignLongestRoad(holder: Party?, candidates: List<LongestRoadCandidate>): Party? {
+
+    val orderedCandidates = candidates.sortedByDescending { it.longestRoadLength }
+
+    // Longest candidate does not have at least 5 roads
+    if (orderedCandidates[0].longestRoadLength < 5)
+        return null
+
+    // Only one player has the longest road
+    if (orderedCandidates[0].longestRoadLength > orderedCandidates[1].longestRoadLength)
+        return orderedCandidates[0].player
+
+    val holderLongestRoadLength = orderedCandidates.first { it.player == holder }.longestRoadLength
+
+    // Holder and another player has same length
+    if (holderLongestRoadLength == orderedCandidates[0].longestRoadLength)
+        return holder
+
+    // All other cases. More than one player has same length and longer than current holder
+    return null
+}
+
+/**
+ * Returns a list with the count of the longest road for each player.
+ */
+fun longestRoadsForAllPlayers(board: PlacedHexTiles,
+                              roads: List<RoadState>,
+                              settlements: List<SettlementState>,
+                              players: List<Party>): List<LongestRoadCandidate> {
+    return players.map { LongestRoadCandidate(it, longestRoadForPlayer(board, roads, settlements, it).count()) }
+}
+
+/**
+ * Returns a set of road ids with the longest road for a particular player.
+ */
+fun longestRoadForPlayer(board: PlacedHexTiles,
+                         roads: List<RoadState>,
+                         settlements: List<SettlementState>,
+                         player: Party): Set<UniqueIdentifier> {
+    val playerRoadIds = roads.filter { it.owner == player }.toSet()
+    val othersSettlements = settlements.filter { it.owner != player }.toSet()
+    return calculateLongestRoad(board, playerRoadIds, othersSettlements)
+}
+
+/**
+ * Calculate the longest road for a player.
+ * roads must only contain roads for the player.
+ * settlements must only contain settlements of all the other players.
+ */
+private fun calculateLongestRoad(board: PlacedHexTiles,
+                                 roads: Set<RoadState>,
+                                 settlements: Set<SettlementState>): Set<UniqueIdentifier> {
+    val roadIds = roads.map { it.linearId }.toSet()
+    val settlementsCorners = settlements.map { AbsoluteCorner(it.hexTileIndex, it.hexTileCoordinate) }.toSet()
+
+    var longestRoad = setOf<UniqueIdentifier>()
+    val neverVisitedRoads = roads.map { it.linearId }.toMutableSet()
+
+    while (neverVisitedRoads.isNotEmpty()) {
+        val firstNeverVisitedRoad = roads.first { it.linearId == neverVisitedRoads.first() }
+        val candidate = calculateLongestRoad(
+                board, AbsoluteSide(firstNeverVisitedRoad.hexTileIndex, firstNeverVisitedRoad.hexTileSide),
+                firstNeverVisitedRoad.linearId, roadIds, settlementsCorners)
```

Comment:
I would pass the whole `firstNeverVisitedRoad` to prevent others from thinking that you can pass values from 2 different roads.

Reply:
I don't understand very well what you mean.
`firstNeverVisitedRoad` is just a road from where to start the algorithm. Tomorrow in the meet up i can explain you the logic behind and it will make sense.

Reply:
If you ask to pass `(absoluteSide, linearId)`, those 2 **may** be not linked. A developer who does not understand could do `(road1.absoluteSide, road2.linearId)`. See? They are not related.
But if you ask to pass `(road)`, then it is forced to be related.

Reply:
I believe @xavierlepretre has a point - then you can extract that data yourself from within the class and perform the necessary checks.